lib/stores/S3Store.js

```diff
@@ -276,57 +276,81 @@ class S3Store extends DataStore {
                 log(`[${metadata.file.id}] finished uploading part #${current_part_number}`);
 
                 return data.ETag;
-            })
-            .catch((err) => {
-                throw err;
             });
     }
 
     /**
-     * Processes the following steps:
-     * - uploads a part (stream) to s3
-     * - finishes the upload in case offset === length
-     *
-     * @param {Object}                metadata            upload metadata
-     * @param {http<IncomingMessage>} req                 incoming request
-     * @param {Number}                current_part_number number of the current part/chunk
-     * @return {Promise<Number>}                         which resolves with the current offset
-     * @memberof S3Store
-     */
-    _processUpload(metadata, req, current_part_number) {
-        return this._uploadPart(metadata, req, current_part_number)
-            .then(() => this.getOffset(metadata.file.id, true))
-            .then((current_offset) => {
-                if (parseInt(metadata.file.upload_length, 10) === current_offset.size) {
-                    return this._finishMultipartUpload(metadata, current_offset.parts)
-                        .then((location) => {
-                            log(`[${metadata.file.id}] finished uploading: ${location}`);
-
-                            this.emit(EVENTS.EVENT_UPLOAD_COMPLETE, {
-                                file: Object.assign({}, metadata.file, { location }),
-                            });
-
-                            this._clearCache(metadata.file.id);
-
-                            return current_offset.size;
-                        })
-                        .catch((err) => {
-                            throw err;
+   * uploads a stream to s3 using multiple parts
+   *
+   * @param {Object}                metadata            upload metadata
+   * @param {http<IncomingMessage>} req                 incoming request
+   * @param {Number}                currentPartNumber number of the current part/chunk
+   * @return {Promise<Number>}                         which resolves with the current offset
+   * @memberof S3Store
+   */
+    _processUpload(metadata, req, currentPartNumber) {
+        return new Promise((resolve, reject) => {
+            const splitter = new FileStreamSplitter(
+                {},
+                { maxChunkSize: this.part_size, directory: '.' }
+            );
+
+            const promises = [];
+
+            let pendingChunkFilepath = null;
+
+            stream.pipeline(req, splitter, (pipelineErr) => {
```

Comment:
Is `stream.pipeline` a better version of the `ReadbleStream#pipe` method?

Reply:
`ReadableStream` is an implementation of the [WHATWG Streams Standard](https://streams.spec.whatwg.org/), currently stage 1 (experimental). It's mostly made to be isomorphic between Node.js and the browser, but it's significantly slower than the older streams API.

Also `pipeline` is a series of streams, wheres `ReadableStream` is a _single_ read stream.

Reply:
Ok, maybe my question was ill-phrased. Let's try again: Why do you use `stream.pipeline` here and not `req.pipe(splitter).on('error', ...)`?

Reply:
I see! `pipeline` is an improved version added later in Node 10 while `pipe` is a lot older` It does a better job at cleaning up if one of the streams failed.

Reply:
Ah good to know, thank you!